engine/lib/analyses/mutable_variables.ml

```diff
@@ -0,0 +1,91 @@
+open Base
+open Utils
+
+module%inlined_contents Make (F : Features.T) = struct
+  module FA = F
+  module A = Ast.Make (F)
+  open Ast
+
+  type pre_data =
+    ( concrete_ident,
+      concrete_ident list,
+      Concrete_ident.comparator_witness )
+    Map.t
+    * (concrete_ident, int, Concrete_ident.comparator_witness) Map.t
+
+  type analysis_data =
+    ( concrete_ident,
+      local_ident list * ((local_ident * A.ty) * int) list,
+      Concrete_ident.comparator_witness )
+    Map.t
+
+  module Flatten = Flatten_ast.Make (F)
+
+  let rec analyse (func_dep : pre_data) (items : A.item list) : analysis_data =
+    let mut_var_list, _ =
+      List.fold_left
+        ~f:(fun y x ->
+          match x.v with
+          | Fn { name; generics; body; params } ->
+              let items, count = analyse_function_body body (snd y) in
+              ([ (name, items) ] @ fst y, count)
+          | _ -> y)
+        ~init:([], 0) items
+    in
+    let mut_map :
+        ( concrete_ident,
+          ((local_ident * A.ty) * int) list,
+          Concrete_ident.comparator_witness )
+        Map.t =
+      List.fold_left
+        ~f:(fun y x -> Map.set y ~key:(fst x) ~data:(snd x))
+        ~init:(Map.empty (module Concrete_ident))
+        mut_var_list
+    in
+    List.fold_left
+      ~f:(fun y x ->
+        Map.set y ~key:x
+          ~data:
+            (let local_muts =
+               match Map.find mut_map x with Some l -> l | None -> []
+             in
+             ( (List.map ~f:(fst >> fst) local_muts
+               @
+               match Map.find (fst func_dep) x with
+               | Some l ->
+                   let l' =
+                     List.concat (List.filter_map ~f:(Map.find mut_map) l)
+                   in
+                   List.map ~f:(fst >> fst) l'
+               | None -> []),
+               local_muts )))
+      ~init:(Map.empty (module Concrete_ident))
+      (List.map ~f:fst mut_var_list)
+
+  and analyse_function_body (x : A.expr) (i : int) :
+      ((local_ident * A.ty) * int) list * int =
+    let mut_var_list =
+      List.dedup_and_sort
+        ~compare:(fun x y -> LocalIdent.compare (fst x) (fst y))
+        (List.filter_map
+           ~f:(fun x ->
+             match x.e with
+             | Assign { lhs = LhsLocalVar { var; typ }; witness } ->
+                 Some (var, typ)
+             | Let { lhs = { p = PBinding { mut = Mutable _; var; typ } }; _ }
+               ->
+                 Some (var, typ)
+             | _ -> None)
+           (Flatten.flatten_ast x))
+    in
```

Comment:
This is exactly a computing a map from `A.expr` to a list of typed ident that were assigned, right?
`Ast_utils.free_assigned_variables` do something a bit different (it's taking care of scoping), but I think you could implement that with a visitor in a very simple way (**without flattening exprs**):
```ml
object
        inherit [_] expr_reduce as super
        inherit [_] Sets.TypedLocalIdent.monoid as m
        method visit_t _ _ = m#zero
        method visit_mutability (_f : unit -> _ -> _) () _ = m#zero
            
        method visit_Assign _env lhs e _wit =
          let rec visit_lhs lhs =
            match lhs with
            | LhsLocalVar { var; _ } ->
                Set.singleton (module TypedLocalIdent) (var, e.typ)
            | LhsFieldAccessor { e; _ } -> visit_lhs e
            | LhsArrayAccessor { e; index; _ } ->
                Set.union (super#visit_expr () index) (visit_lhs e)
            | LhsArbitraryExpr { witness; e } -> m#zero
          in
          visit_lhs lhs
      end
```

(note that this codes handles all kinds of LHSs)